openid-4-verifiable-presentations-1_0.md

```diff
@@ -616,11 +622,310 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+query. This specification does not define a specific structure or specific values
+for this property. The purpose is intended to be used by the Verifier to
+communicate the reason for the query to the Wallet. The Wallet MAY use this
+information to show the user the reason for the request.
+
+## Claims Query {#claims_query}
+
+Each entry in `claims` MUST be an object with the following properties:
+
+`id`:
+: REQUIRED if `claim_sets` is present in the Credential Query; OPTIONAL otherwise. A string
+identifying the particular claim. The value MUST be a non-empty string
+consisting of alphanumeric, underscore (`_`) or hyphen (`-`) characters.
+Within the particular `claims` array, the same `id` MUST NOT
+be present more than once.
+
+`path`:
+: REQUIRED if the Credential Format uses a JSON-based claims structure; MUST NOT
+be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
+within the Verifiable Credential, as defined in (#claims_path_pointer).
+
+`namespace`:
+: REQUIRED if the Credential Format is based on the mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the namespace of the data element
+within the mdoc, e.g., `org.iso.18013.5.1`.
+
+`claim_name`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the data element identifier of the data element within the provided namespace
+in the mdoc, e.g., `first_name`.
+
+`values`:
+: OPTIONAL. An array of strings, integers or boolean values that specifies the expected values of the claim.
+If the `values` property is present, the Wallet SHOULD return the claim only if the
+type and value of the claim both match for at least one of the elements in the array. Details of the processing rules are defined in (#selecting_claims).
+
+### Selecting Claims and Credentials {#vp_query_lang_processing_rules}
+
+The following section describes the logic that applies for selecting claims 
+and for selecting credentials. 
+
+Note: While this specification provides the mechanisms for requesting different
+sets of claims and credentials, it does not make assumptions about the user 
+interface of the Wallet, for example, if users can select which claims to present 
+or which combination of credentials to present.
+
+#### Selecting Claims {#selecting_claims}
+
+The following rules apply for selecting claims via `claims` and `claim_sets`:
+
+- If `claims` is not provided, the Verifier requests all claims existing
+  in the Credential.
+- If `claims` is provided, but `claim_sets` is not provided,
+  the Verifier requests all claims listed in `claims`.
+- Otherwise, the Verifier requests one combination of the claims listed in
+  `claim_sets`. The order of the options conveyed in the `claim_sets`
+  array expresses the Verifiers' preference for what is returned; the Wallet MUST return
```

Comment:
```suggestion
  array expresses the Verifiers' preference for what is returned; the Wallet SHOULD return
```
as per @martijnharing 's comment, the RP cannot mandate the wallet to do anything.

Reply:
That doesn't mean we cannot define rules for the Wallet. There may be other reasons to say "SHOULD" here and I'm open to discuss them, but this is not a good reason.

Reply:
Also strongly in favor in making this a SHOULD as describes in my comment earlier. Since from an RPs perspective all the options in the array should be sufficient to achieve the use case, there is no reason to mandate a particular option.

Reply:
If we use the word MUST then a conformance test will be written for the wallet, and wallets that allow the user to choose a low priority set will fail the test. This is an important consideration for the EDIW. Do we really want wallets that allow user choice to fail to be certified?

Reply:
If we define that there is no difference to the RP which of the elements in claim_sets is returned means that there is no necessity to have a user choice and allows the wallet to return whatever is simplest. It makes it explicit that it doesn't matter what in the array is returned but also makes it clear that in the absence of any other wallet logic it should just be the first element. Therefore I'm in favor of recommending to return the first item in the order of preference.
However, actually mandating behavior on the wallet side means that there is no leeway to the wallet to deviate for any reason. Some examples where the wallet may want to deviate from the mandatory behavior:

- Request for [age_over_18, date_of_birth], the wallet wants to return age_over_21.
- Request for [on_device_auth, portrait], where the on_device_auth element requires an external factor that is not available, or adds significant friction that the user does not want to accept.
- Request for [element_y, element_x] and [element_a, element_x] where the credential contains element_x and element_y. Returning the first element element of the array would result in returning more than required.

If the only reason to actually mandate the behavior is because it would address the 'conditional' use-case, that to me does not seem to be worth the potential issues caused by that. As you mentioned, conditional is likely a very specific set of use cases and making I specific mechanism for that therefore seems fitting. We could also explore whether addressing it through issuer selection can address all the use cases otherwise handled by 'conditional'.

Reply:
> * Request for [age_over_18, date_of_birth], the wallet wants to return age_over_21.

This would be unexpected for the Verifier and hard to process. If the verifier knows age_over_21, with the current solution, it could just list it first.

> * Request for [on_device_auth, portrait], where the on_device_auth element requires an external factor that is not available, or adds significant friction that the user cannot accept at that time.

In this case, the claim might just not be available - we can add more precise language around what "unavailable" means later on.

> * Request for [element_y, element_x] and [element_a, element_x] where the credential contains element_x and element_y. Returning the first element element of the array would result in returning more than required.

I don't understand. If the credential contains both y and x, it can just send it.

Reply:
Regarding "purpose", each claim set in the claim sets should have a purpose so that the user knows what the purpose of the claim set is. It does not mater what each alternative choice in the claim set is, because they should all be equivalent from the RP's perspective. So I maintain that the user should be able to choose between the alternative choices in a claim set, and either go with the RP's preferred priority, or not.